Model User and simple controller(Feature/32)

Declow0 opened the pull request:
First realization of User model. Create POJO User class. You can get user JSON objest by POST request on "/user" end-point .

Card: #32

Declow0 committed:
Test version of User model

Card: #32

Declow0 committed:
Change field eMail to email, add exeptions

Card: #32

Declow0 commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Как разработчик фронтенда, хочу сущность "Пользователь", для того чтобы назначать на себя задачи, а также комментировать другие задачи под своим именем](https://trello.com/c/iJFr9vUm/32--)

Declow0 committed:
Add EOF at the end of file

Card: #32

kevromster commented on src/main/java/edu/xored/tracker/User.java:
```diff
@@ -0,0 +1,31 @@
+    private String email;
+
+    public User() {
+        this(null, null);
```
этот _this()_ не нужен, т.к. не несёт никакой смысловой нагрузки

Declow0 replied:
Изменен на конструктор без тела. Done.

kevromster commented on src/main/java/edu/xored/tracker/UserController.java:
```diff
@@ -0,0 +1,56 @@
+import org.springframework.web.bind.annotation.ResponseStatus;
+import org.springframework.web.bind.annotation.RestController;
+
+import java.io.*;
```
вместо звёздочек - явные импорты того, что необходимо

Declow0 replied:
Done.

kevromster commented on src/main/java/edu/xored/tracker/UserController.java:
```diff
@@ -0,0 +1,56 @@
+    }
+
+    @ResponseStatus(value = HttpStatus.FAILED_DEPENDENCY, reason = "Error on inStream.readLine()")
+    private class InSrteamException extends RuntimeException {
```
Очепятка в имени класса. В целом - название не говорящее и завязанное на специфику места использования. Лучше придумать что-то другое.
reason тоже не очень написан - а если я там код изменю с readLine на что-то другое?

Declow0 replied:
Поменял на OutputException. Вроде говорящее и использовать можно не только тут. Ну и сделать комментарий TODO про Exceptions как говорили на митинге.
Done.

kevromster commented on src/main/java/edu/xored/tracker/UserController.java:
```diff
@@ -0,0 +1,56 @@
+        String name = null;
+        String email = null;
+        try {
+            theProcess = Runtime.getRuntime().exec("git config user.name");
```
1) Код получения theProcess здесь и ниже повторяется (со всеми try/catch). Копипаста зло, нужно выделить отдельный метод.
2) Имена команд "git config blabla..." лучше не хардкодить прямо тут, а вынести в отдельную private static final String константу класса.

Declow0 replied:
1) Создан метод getOutput().
2) А зачем final?  Ведь все инициализированные строки вроде по умолчанию final... Или я что-то неправильно понимаю...
Done.

kevromster commented on src/main/java/edu/xored/tracker/UserController.java:
```diff
@@ -0,0 +1,56 @@
+            throw new GitNotFoundException();
+        }
+        try {
+            inStream = new BufferedReader(new InputStreamReader( theProcess.getInputStream() ));
```
1) тут бы лучше try-with-resources конструкцию применить, почитай про неё
2) опять же код для email ниже повторяется

Declow0 replied:
1) Реализовано. Узнал,что данная возможность появилась с появлением Java 7.
2) Исправлено выше.
Done.

Declow0 committed:
Change default constructor

Default constructor doesn't have body

Card: #32

Declow0 committed:
Fixed bugs in code

Change import, now without java.io.*. Add method getOutput() to get data from git command. Add 2 constants of commands. Now for work with buffer using try-with-resources. Change exception's description for Output. Add TODO comment.

Card: #32

Declow0 committed:
Now end-point of controller to /api/user

Card: #32

kevromster commented on src/main/java/edu/xored/tracker/UserController.java:
```diff
@@ -0,0 +1,50 @@
+@RestController
+@RequestMapping(value = "/api")
+public class UserController {
+    private static String getNameCommand = "git config user.name";
```
здесь нужен `final` для того, чтобы нельзя было сделать так:

```
private void someMethod() {
  getNameCommand = "another command";
}
```

Declow0 replied:
Done.

kevromster commented on src/main/java/edu/xored/tracker/UserController.java:
```diff
@@ -0,0 +1,50 @@
+    private static String getNameCommand = "git config user.name";
+    private static String getEmailCommand = "git config user.email";
+
+    String getOutput (String command) {
```
private?

Declow0 replied:
Done.

kevromster commented on src/main/java/edu/xored/tracker/UserController.java:
```diff
@@ -0,0 +1,50 @@
+
+    String getOutput (String command) {
+        Process theProcess;
+        // TODO: How to work with exceptions? How to log them?
```
давай перенесём этот `TODO` в `catch` рядом с `e.printStackTrace();` и добавим номер задачи, вроде такого:

```
// TODO (Card #50): find better approach for exceptions logging
```

Declow0 replied:
Done.

kevromster commented on src/main/java/edu/xored/tracker/UserController.java:
```diff
@@ -0,0 +1,50 @@
+            e.printStackTrace();
+            throw new GitNotFoundException();
+        }
+        try (BufferedReader inStream = new BufferedReader(new InputStreamReader( theProcess.getInputStream() ))) {
```
а зачем тут дополнительные пробелы внутри скобок `()`? кажется, не нужны

Declow0 replied:
Done.

kevromster commented on src/main/java/edu/xored/tracker/UserController.java:
```diff
@@ -0,0 +1,50 @@
+    }
+
+    @ResponseStatus(value = HttpStatus.FAILED_DEPENDENCY, reason = "An error occurred while receiving the output")
+    private class ReceiveOutputException extends RuntimeException {
```
давай назовём его `ExecutionFailedException`, и reason напишем вроде "ошибка выполнения команды"

Declow0 replied:
Done.

Declow0 committed:
Fixed bugs in code

Change name from OutputExecption to ExecutionFailed and text of this exception. Commands are final now. Change position and text of TODO. Method getOutput is private now.

Card: #32

kevromster commented on src/main/java/edu/xored/tracker/UserController.java:
```diff
@@ -0,0 +1,51 @@
+    private static final String getNameCommand = "git config user.name";
+    private static final String getEmailCommand = "git config user.email";
+
+    private String getOutput (String command) {
```
Обычно private-методы после public идут. Надо будет это в style guide прописать. Давай передвинем. Остальное всё ок.

Declow0 replied:
Done.

Declow0 committed:
Change position of  methods. At first public after private

Card: #32
